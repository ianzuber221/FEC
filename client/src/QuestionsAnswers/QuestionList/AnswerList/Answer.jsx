import React, { useState } from 'react';
import { format } from 'date-fns';
import styled from 'styled-components';
import Helpful from '../../components/Helpful';
import Report from '../../components/Report';
import Image from '../../components/Image';

function Answer({ answer, renderAnswers, last }) {
  const [helpfulCount, setHelpfulCount] = useState(answer.helpfulness)
  const answerInfo = (
    <span>
      {answer.answerer_name.toLowerCase() === 'seller' ? <span>by <b>Seller</b></span> : <span>by {answer.answerer_name} </span>}
      {', '}
      {format(new Date(answer.date), 'MMMM d, yyyy')}
    </span>
  );
  const images = answer.photos.length > 0 && (
    <PhotoDiv>
      {answer.photos.map((photo, index) => (
        <Image
          key={index}
          url={typeof photo === 'object' ? photo.url : photo}
        />
      ))}
    </PhotoDiv>
  );
  const helpful = (
    <Helpful id={answer.answer_id || answer.id}
      type="answer"
      currentCount={helpfulCount}
      setHelpfulCount={setHelpfulCount}
    />
  );
  const report = (
    <span>
      <Report
        id={answer.answer_id || answer.id}
        type="answer"
      />
    </span>
  );

  return (
    <AnswerDiv className="answer" last={last}>
      <div className="answer-text">{answer.body}</div>
      {images}
      <OptionsDiv>
        {answerInfo}
        <PaddedSpan>{' | '}</PaddedSpan>
        {helpful}
        <PaddedSpan>{' | '}</PaddedSpan>
        {report}
      </OptionsDiv>
    </AnswerDiv>
  );
}

export default Answer;

const PhotoDiv = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  width: 100%;
  overflow-x: auto;
  margin: 5px 0;
`

const PaddedSpan = styled.div`
  padding: 0 10px;
  font-size: 15px;
`;

const OptionsDiv = styled.div`
  font-size: 11px;
  font-weight: 300;
  color: #77787a;
  display: flex;
  max-width: 40%;
  padding: 10px 0;
  align-items: center;
`;

const AnswerDiv = styled.div`
  padding-bottom: ${props => props.last ? 0 : '5px'};
  border-bottom: ${props => props.last ? 'none' : '0.5px solid grey;'};
  margin-bottom: ${props => props.last ? 0 : '15px'};
`;