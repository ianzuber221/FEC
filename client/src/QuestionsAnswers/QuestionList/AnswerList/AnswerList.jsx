import React, { useState } from 'react';
import styled from 'styled-components';
import Answer from './Answer';

function AnswerList({ answerList, renderAnswers, Title }) {
  const [collapsed, setCollapsed] = useState(true);
  const filteredAnswerList = collapsed ? answerList.slice(0, 2) : answerList;
  if (Array.isArray(answerList) && answerList.length > 0) {
    return answerList.length > 0 && (
      <DivAnswer>
        <Title>A:</Title>
        <AnswerListDiv>
          {collapsed
          && (
            <div>
              {filteredAnswerList
                .map((answer, index) => (
                  <Answer
                    key={answer.id || answer.answer_id}
                    answer={answer}
                    renderAnswers={renderAnswers}
                    last={index === filteredAnswerList.length - 1}
                  />
                ))}
            </div>
          )}
          {!collapsed
          && (
            <Scroller id="scrollableDiv">
              {filteredAnswerList.map((answer, index) => (
                <Answer
                  key={answer.id || answer.answer_id}
                  answer={answer}
                  renderAnswers={renderAnswers}
                  last={index === filteredAnswerList.length - 1}
                />
              ))}
            </Scroller>
          )}
          {collapsed && answerList.length > 2
          && <CollapseSpan onClick={() => setCollapsed(false)}>LOAD MORE ANSWERS</CollapseSpan>}
          {!collapsed
          && <CollapseSpan onClick={() => setCollapsed(true)}>COLLAPSE ANSWERS</CollapseSpan>}
        </AnswerListDiv>
      </DivAnswer>
    );
  }
  return (
    <DivAnswer>
      <Title />
      <div>No answers.</div>
    </DivAnswer>
  );
}

export default AnswerList;

const CollapseSpan = styled.span`
  font-weight: bold;
  font-size: 11px;
`;

const DivAnswer = styled.div`
  display: flex;
  margin-bottom: 20px;
  font-size: small;
`;

const Scroller = styled.div`
  max-height: 50vh;
  overflow-y: auto;
  height: 300;
  padding-right: 10px;
  display: 'flex';
  flex-direction: 'column-reverse';
`;

const AnswerListDiv = styled.div`
  width: 95%;
`;
