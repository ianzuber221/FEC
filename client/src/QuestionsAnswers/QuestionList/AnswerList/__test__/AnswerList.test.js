/**
 * @jest-environment jsdom
 */
/* eslint-disable react/jsx-filename-extension */

import React from 'react';
import renderer from 'react-test-renderer';
import { render } from '@testing-library/react';
import styled from 'styled-components';
import userEvent from '@testing-library/user-event';
import { jest } from '@jest/globals';
import AnswerList from '../AnswerList';
import '@testing-library/jest-dom/extend-expect';

const renderMock = jest.fn();

const answerList = [
  {
    id: 123,
    answerer_name: 'test',
    date: new Date('December 17, 1995 03:24:00'),
    photos: [],
    body: 'new answer',
  },
  {
    id: 124,
    answerer_name: 'test',
    date: new Date('December 17, 1995 03:24:00'),
    photos: [],
    body: 'new answer',
  },
  {
    id: 125,
    answerer_name: 'test',
    date: new Date('December 17, 1995 03:24:00'),
    photos: [],
    body: 'new answer',
  },
];

const Title = styled.div`
`;

it('should match snapshot', () => {
  const component = renderer.create(
    <AnswerList
      answerList={answerList}
      renderAnswers={renderMock}
      Title={Title}
    />
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('when clicked "LOAD MORE ANSWERS", text should change to "COLLAPSE ANSWERS"', async () => {
  const { getByText } = render(
    <AnswerList
    answerList={answerList}
    renderAnswers={renderMock}
    Title={Title}
    />
  );

  const load = getByText('LOAD MORE ANSWERS');
  expect(load).toBeInTheDocument();
  await userEvent.click(load);
  expect(getByText('COLLAPSE ANSWERS')).toBeInTheDocument();
})