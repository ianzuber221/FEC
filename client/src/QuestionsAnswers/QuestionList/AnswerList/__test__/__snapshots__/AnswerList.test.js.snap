// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<div
  className="sc-jqUVSM kYeeqJ"
>
  <div
    className="sc-crXcEl ikVVYA"
  >
    A:
  </div>
  <div
    className="sc-iqcoie lbKltb"
  >
    <div
      className="sc-kDDrLX jXQQSG"
      id="scrollableDiv"
    >
      <div
        className="sc-iBkjds cdgNPz answer"
      >
        <div
          className="answer-text"
        >
          new answer
        </div>
        <div
          className="css-1iivxpu-MuiStack-root"
        >
          <span>
            <span>
              by 
              test
               
            </span>
            , 
            December 17, 1995
          </span>
          <hr
            className="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical MuiDivider-flexItem css-18i4s2i-MuiDivider-root"
          />
          <div
            className="sc-dkzDqf hoZXih helpful"
          >
            <span
              className="sc-bczRLJ grWmXc"
            >
              Helpful?
            </span>
            <u
              className="sc-gsnTZi hBfuwO"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="button"
              tabIndex="-1"
            >
              Yes
            </u>
            <span>
              (undefined)
            </span>
          </div>
          <hr
            className="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical MuiDivider-flexItem css-18i4s2i-MuiDivider-root"
          />
          <span>
            <u
              onClick={[Function]}
              onKeyDown={[Function]}
              role="button"
              tabIndex="-1"
            >
              Report
            </u>
          </span>
        </div>
      </div>
      <div
        className="sc-iBkjds cdgNPz answer"
      >
        <div
          className="answer-text"
        >
          new answer
        </div>
        <div
          className="css-1iivxpu-MuiStack-root"
        >
          <span>
            <span>
              by 
              test
               
            </span>
            , 
            December 17, 1995
          </span>
          <hr
            className="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical MuiDivider-flexItem css-18i4s2i-MuiDivider-root"
          />
          <div
            className="sc-dkzDqf hoZXih helpful"
          >
            <span
              className="sc-bczRLJ grWmXc"
            >
              Helpful?
            </span>
            <u
              className="sc-gsnTZi hBfuwO"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="button"
              tabIndex="-1"
            >
              Yes
            </u>
            <span>
              (undefined)
            </span>
          </div>
          <hr
            className="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical MuiDivider-flexItem css-18i4s2i-MuiDivider-root"
          />
          <span>
            <u
              onClick={[Function]}
              onKeyDown={[Function]}
              role="button"
              tabIndex="-1"
            >
              Report
            </u>
          </span>
        </div>
      </div>
    </div>
    <span
      className="sc-papXJ ceIjkY"
      onClick={[Function]}
    >
      LOAD MORE ANSWERS
    </span>
  </div>
</div>
`;
